2007-04-13  Jakub Jelinek  <jakub@redhat.com>

        * stdio-common/printf_fp.c (___printf_fp): Fix exponent -4
        special case handling when wcp == wstartp + 1.

--- libc/stdio-common/printf_fp.c.jj    2007-02-26 18:13:47.000000000 +0100
+++ libc/stdio-common/printf_fp.c       2007-04-13 16:03:38.000000000 +0200
@@ -1081,12 +1081,18 @@
 	    /* This is another special case.  The exponent of the number is
 	       really smaller than -4, which requires the 'e'/'E' format.
 	       But after rounding the number has an exponent of -4.  */
-	    assert (wcp >= wstartp + 2);
+	    assert (wcp >= wstartp + 1);
 	    assert (wstartp[0] == L'1');
 	    __wmemcpy (wstartp, L"0.0001", 6);
 	    wstartp[1] = decimalwc;
 	    wmemset (wstartp + 6, L'0', wcp - (wstartp + 2));
-	    wcp += 4;
+	    if (wcp >= wstartp + 2)
+	      {
+		wmemset (wstartp + 6, L'0', wcp - (wstartp + 2));
+		wcp += 4;
+	      }
+	    else
+	      wcp += 5;
 	  }
 	else
 	  {
